File: Tree/99. Recover Binary Search Tree/99. Recover Binary Search Tree.java
class Solution {
    TreeNode first = null;
    TreeNode second = null;
    TreeNode lastSeen = null;
    public void recoverTree(TreeNode root) {
        inorder(root);
        int temp = second.val;
        second.val = first.val;
        first.val = temp;
    }
    private void inorder(TreeNode root) {
        if (root == null) return;
        inorder(root.left);
        // .....
        if (lastSeen == null || root.val > lastSeen.val) {
            lastSeen = root;
        } else {
            if (first == null) {
                first = lastSeen;
                lastSeen = root;
                second = root;
            } else {
                second = root;
                return;
            }
        }
        inorder(root.right);
    }
}

File: Tree/99. Recover Binary Search Tree/99. Recover Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    TreeNode* first = NULL;
    TreeNode* second = NULL;
    TreeNode* lastSeen = new TreeNode(INT_MIN);
public:
    void recoverTree(TreeNode* root) {
        inorder(root);
        int temp = first->val;
        first->val = second->val;
        second->val = temp;
    }
    void inorder(TreeNode* root) {
        // corner case
        if (root == NULL) return;
        inorder(root->left);
        // ....
        if (root->val < lastSeen->val) {
            if (first == NULL) {
                first = lastSeen;
            }
            second = root;
        }
        lastSeen = root;
        inorder(root->right);
    }
};
